### PR TITLE
Use bytes / bytes_view to handle buffers

### PR DESCRIPTION
**Pull request by chfast**

### CODE REVIEW

**Comment by axic** [comment]
Can you split the CI stuff into a new PR?

**Comment by axic** [comment]
Do you see any speed difference? Or is it just code clarity at the expense of needing C++17?

**Comment by chfast** [comment]
No speed was measured.

**Comment by chfast** [comment]
No speed was measured.

File: src/eei.h [hunk 1]
```diff
@@ -53,13 +58,13 @@ class EthereumInterface {
 public:
   explicit EthereumInterface(
     evmc_context* _context,
-    std::vector<uint8_t> const& _code,
+    std::vector<uint8_t> const& _code,  // TODO: Change arg type to bytes or bytes_view.
```
Comment (axic) [inline src/eei.h]: Changing this requires a lot of work or what is the reason postponing it?

File: src/binaryen.cpp [hunk 2]
```diff
@@ -509,10 +509,12 @@ ExecutionResult BinaryenEngine::execute(
...
 }
 
-void BinaryenEngine::loadModule(vector<uint8_t> const& code, wasm::Module & module)
+void BinaryenEngine::loadModule(bytes_view code, wasm::Module& module)
 {
   try {
-    wasm::WasmBinaryBuilder parser(module, reinterpret_cast<vector<char> const&>(code), false);
+    auto const b = reinterpret_cast<char const*>(code.begin());
+    auto const e = reinterpret_cast<char const*>(code.end());
+    wasm::WasmBinaryBuilder parser(module, {b, e}, false);
```
Comment (axic) [inline src/binaryen.cpp]: These three lines are way less clear to me compared to the original one.

File: src/eei.h [hunk 3]
```diff
@@ -24,6 +26,9 @@
 
 namespace hera {
 
+using bytes = std::basic_string<uint8_t>;
+using bytes_view = std::basic_string_view<uint8_t>;
```
Comment (axic) [inline src/eei.h]: Moving these to `helpers.h` may enable sharing them across the code.

File: src/binaryen.cpp (hunk 2 above)
Comment (chfast) [reply src/binaryen.cpp]: At least vector is not casted.

**Comment by chfast** [comment]
> The benefit here is mostly readability?

No. Compatibility, i.e. you can easily convert any `uint8_t` buffer (vector, C array, string, etc) to `bytes_view`.

The build with libc++ is failing because it does not like `basic_fstream<uint8_t>`. I invented this on the fly, have to check if that's compliant with C++.

File: src/hera.cpp [hunk 4]
```diff
@@ -131,19 +131,19 @@ pair<evmc_status_code, vector<uint8_t>> callSystemContract(
   if (result.release)
     result.release(&result);
 
-  return make_pair(result.status_code, ret);
+  return {result.status_code, ret};
```
Comment (axic) [inline src/hera.cpp]: You want to merge such simplifications separately?

File: src/hera.cpp [hunk 5]
```diff
@@ -273,12 +273,12 @@ evmc_result hera_execute(
     heraAssert(hera->engine, "Wasm engine not set.");
     WasmEngine& engine = *hera->engine;
 
-    ExecutionResult result = engine.execute(context, run_code, state_code, *msg, meterInterfaceGas);
+    ExecutionResult result = engine.execute(context, {run_code.data(), run_code.size()}, state_code, *msg, meterInterfaceGas);
```
Comment (axic) [inline src/hera.cpp]: Is there no nicer way to do this? Cannot one pass `run_code` directly?

File: src/eei.cpp [hunk 6]
```diff
@@ -246,8 +245,9 @@ bool exceedsUint128(evmc_uint256be const& value) noexcept
       safeChargeDataCopy(length, GasSchedule::extcode);
 
       evmc_address address = loadAddress(addressOffset);
-      // TODO: optimise this so no vector needs to be created
-      vector<uint8_t> codeBuffer(length);
+      // TODO: optimise this so no copy needs to be created
+      bytes codeBuffer;
+      codeBuffer.resize(length);
```
Comment (axic) [inline src/eei.cpp]: I see below you use `data(length, 0)` to zero-fill. Can you do the same here?

File: src/binaryen.cpp [hunk 7]
```diff
@@ -509,10 +509,13 @@ ExecutionResult BinaryenEngine::execute(
...
 }
 
-void BinaryenEngine::loadModule(vector<uint8_t> const& code, wasm::Module & module)
+void BinaryenEngine::loadModule(bytes_view code, wasm::Module& module)
 {
   try {
-    wasm::WasmBinaryBuilder parser(module, reinterpret_cast<vector<char> const&>(code), false);
+    auto const b = reinterpret_cast<char const*>(code.begin());
+    auto const e = reinterpret_cast<char const*>(code.end());
+    std::vector<char> codeCopy{b, e};
```
Comment (axic) [inline src/binaryen.cpp]: Is there no way to avoid the copy here just because change the type form `uint8_t` to `char`?

Comment (chfast) [reply src/binaryen.cpp]: No other than improving Binaryen API.

**Comment by axic** [comment]
Ah, this actually passes tests just codecov is complaining.

File: src/eei.h [hunk 8]
```diff
@@ -201,9 +206,9 @@ class EthereumInterface {
 
   evmc_tx_context m_tx_context{};
   evmc_context* m_context = nullptr;
-  std::vector<uint8_t> const& m_code;
+  bytes m_code;
```
Comment (axic) [inline src/eei.h]: This should be `bytes_view`.

File: src/binaryen.cpp (hunk 7 above)
Comment (axic) [reply src/binaryen.cpp]: Improving how? Binaryen already takes a reference.

File: src/eei.cpp [hunk 9]
```diff
@@ -267,8 +266,8 @@ void WasmEngine::collectBenchmarkingData()
       safeChargeDataCopy(length, GasSchedule::extcode);
 
       evmc_address address = loadAddress(addressOffset);
-      // TODO: optimise this so no vector needs to be created
-      vector<uint8_t> codeBuffer(length);
+      // TODO: optimise this so no copy needs to be created
+      bytes codeBuffer(length, 0);
```
Comment (axic) [inline src/eei.cpp]: What is the `0` argument?

File: src/eei.h [hunk 10]
```diff
@@ -83,13 +86,13 @@ class EthereumInterface {
...
-    std::vector<uint8_t> const& _code,
+    bytes_view _code,
     evmc_message const& _msg,
-    ExecutionResult & _result,
+    ExecutionResult& _result,
     bool _meterGas
   ):
     m_context(_context),
-    m_code(_code),
+    m_code{_code.begin(), _code.end()},
```
Comment (axic) [inline src/eei.h]: Is this a copy here?

**Comment by axic** [comment]
@chfast collapsed the outdated comments, but after a review there are still unanswered questions. Can you please take a look?

File: src/binaryen.cpp (hunk 7 above)
Comment (chfast) [reply src/binaryen.cpp]: The problem is Binaryen forces you to pass a vector.

**Pull request merged**